### DISCUSSION TITLE
Contributor meeting September 2024

### DISCUSSION THREAD

**Discussion by Aggror** [stride3d/stride#2353] [Contributors]
It is time for another Stride engine contributor meeting (purely for the Stride repo contributors for now to keep clutter to a minimum)
This could be a Discord meeting. Since time zone is a difficulty, we can also just use a dedicated discord channel for contributors,

I will update the list as we go.

Topics (WIP)

- Open collective projects semi finished PR
- Payout: https://opencollective.com/stride3d/projects/fbx-import-cpp-to-c-sharp
- Payout? https://opencollective.com/stride3d/projects/multiple-animations-per-model
- Status Morph targets PR:  https://opencollective.com/stride3d/projects/morph-targets/
- Things being worked on
- Website changes
- Docs changes
- Releasing

**Meeting Summary**

1. **Docs - 2D Tutorial:** We can go ahead once Bepu is out and stable. Emanudev/EmanualGF expressed positive interested 
1. **Infrastructure:** WIP, we should speak with .NET Foundation or check other projects regarding the Privacy Policy
1. **Twitter vs X vs Mastodon:** This needs to be addressed by @tebjan and @Aggror 
1. **Game promotion:** Agreed, in the form of [2 blog posts](https://discord.com/channels/500285081265635328/500292618278928385/1284547241008169124)
1. **Website, Docs - Update platform support:** https://github.com/stride3d/stride-website/pull/309
1. **Research:** This doesn't make a sense, because we couldn't act based on this feedback, because we are a small team who is working on the  parts we are passionate about and we can't commit to something we don't have skills even though it might have lots of votes
1. **.NET 9:** We should go ahead with beta branch, prototyping and testing, not necessarily creating NuGet packages
1. **Stride NuGet Packages:** This is already taken care of. There are some old dependencies, which we might not be able to remove, till we switch to Avalonia
1. **xen2, manio and Joreyk:** They are tackling Roslyn generator subject
1. **Bepu integration:** 
   - We will keep both versions for some time and decide later, once Bepu is stable
   - Dobrez is working on Bepu Navigation
1. **Avalonia integration:** Kryptos is still working on the big PR, so others could join him once this is merged

**Comment by VaclavElias** [comment]
**Subjects to discuss:** 
- **Docs - Tutorial:** Can we ask https://github.com/EmanuelGF to create for us a full 2D game tutorial and allocate some budget? He already created a few nice tutorials here https://www.youtube.com/watch?v=XiPzWupYEI0
  - Collaborate with Aggor
- **Infrastructure:** Check with @xen2, can we delete all metrics (database), or keep some rough data. 
  - Related to https://github.com/stride3d/stride/pull/2261
- **Website**: Privacy Policy page update - if we remove metrics, can we remove related matters from the privacy policy?
- **Website**: Marketing: Twitter or X or Mastodon https://github.com/stride3d/stride-website/pull/305
  - Also, I wouldn't be rushing changing Twitter to X, observing that also Unity and Godo still keep Twitter logo
  - If I am correct, we can't share/manage one Mastodon account. We would need to use another 3rt party application, where multiple people can post to this one account.
- **Website**:  Game promotion request - https://discord.com/channels/500285081265635328/500292618278928385/1270413406742253620
  - How can we help? Shall we allow a guest blog post?
  - From greedisgood210
    - About the promotion thing on your website, already asked it but didn't get feedback 🙂 
    - Do you have any wishes on sponsoring the engine itself (pre-trailer animation, game splashscreen etc.). If you have the material let me know and I can add it.
- **Website, Docs [Important]** : https://github.com/stride3d/stride-website/issues/308, @xen2 any input
- **Research**: Let's do 2024 Survey and repeat every year? Draft here: https://docs.google.com/forms/d/e/1FAIpQLSdPtUeRHi4ktpY3AhyYElyN9YzeeF_umoPznkQ4qGqSfb6WyQ/viewform, provided [by OURABIG](https://discord.com/channels/500285081265635328/500285081265635330/1255166894042906705).
- **New Release**: Community is going to be curious about 4.3 .NET 9
- **Stride: NuGet Packages:** We are referencing vulnarable NuGet packages, shall we upgrade all NuGet packages (if possible) so we are using to the newest versions
  - Example, `NuGet.Protocol` in `Stride.Core.Packages` project is referencing vulnarable `NuGet.Packaging 6.8.0`
  - In Visual Studio, click **Show only vulnarable**, hover a mouse over the package and see the dependencies
  ![image](https://github.com/user-attachments/assets/97313dd6-0d8b-41ab-af5b-7500307845b2)


**What I am working on:**
- I am currently writing a huge blog post [Stride Community Toolkit Preview - Code-Only Feature - Basics](https://github.com/VaclavElias/vaclavelias.github.io/blob/post-toolkit/vaclavelias.github.io/posts/2024-08-30-stride-community-toolkit-preview-code-only-feature-basics.md). Would you like it to be posted on Stride Blog or I can do the same but a much shorter version for Stride Blog.
- I’m still working on  Stride Community Toolkit, mainly refactoring and organizing things, using NDepend to review and improve code
  - I'm also adding more examples, particularly trying to figure out 2D examples
  - Looking forward to adding Bepu examples
- I would like to improve Stride Docs - [Contributing](https://doc.stride3d.net/latest/en/contributors/index.html) section, 
  - Improve content consistency
  - Add more content
  - Add chart dependencies from NDepend
  - Add more content related [NDepend report](https://doc.stride3d.net/latest/en/contributors/engine/static-analysis.html)
- Keeping an eye on our site static generators **docfx** and **11ty** and updating accordingly

**Comment by IXLLEGACYIXL** [comment]
topic,, how far should we go with roslyn

**Reply by Jklawreszuk** [reply]
I'll throw in a couple of threads we can touch on 😄 : linux, editor, graphics (drop support for dx 9, other topics)

**Reply by VaclavElias** [reply]
You mean that roslyn should touch these? Or you meant to comment at the bottom to open a separate thread? 🙂

**Reply by IXLLEGACYIXL** [reply]
a topic about roslyn source generation as the conversion from assembly processor is possible BUT it wil lresult in breaking changes 
so currently its unclear which parts should stay in mono.cecil and which parts should be rewritten to roslyn

how to deal with the situation and the breaking changes... atopic which is unclear 

i can convert it to roslyn but core contributors should decide what should be rewritten and how it should be designed

**Reply by Eideren** [reply]
If there's one topic we have to address right now, it would be this one. Could you amend your first post here by summarizing what this is about, what we're trying to solve, what solution we could look into, how it's going on your end / what's blocking you, if you need any help or insight ?
That way other contributors would be up to date on this one, and we can get the ball rolling during the meeting

**Reply by IXLLEGACYIXL** [reply]
source generation works only with incremental source generators , these are syntax based and not on semantic

so datacontract inherited is semantic based and if we source gen on semantic based stuff it just doesnt work, VS cant open anymore stride

if we want source generation it would be required to swap from semantic to syntactical so datacontract attribute on each class instead of inherted bool, this would break everything , you can make quick fixers which are semantic based so if a class inherits from true inherited datacontract i can make a quick fixer to add a datacontract automatically

so if we swap one click to "fix solution" and everywhere is a datacontract where needed and a new class would get the quick fix for the datacontract

for assemblyscan this can also be applied and with that we can make serialziers and assemblyscanners source generated and half of assemblyprocessor is gone also

the bad thing is each dev has to click the quick fix on each class, we can create templates and such to reduce it but it generates overhead

to do it on compilation only is not recommended by fred who made the incremental source gen as the benefit of source gen is to see live the generation

when the serializers are in roslyn yaml + object descriptor can be moved to roslyn as well, without the swap to syntax based attributes yaml and object descriptors cant be converted to roslyn

**Reply by IXLLEGACYIXL** [reply]
im currently not doing anything as i dont know how this roslyn issue will be handled if its clear i will continue in the way the comunity wants it

**Reply by manio143** [reply]
@IXLLEGACYIXL I was having a look around and found this `RegisterImplementationSourceOutput` https://github.com/dotnet/roslyn/pull/54798 which kind of fits into our use case. Now, they say that it's not yet handled by the IDE the right way, but I don't think the cost of doing work in our generator will be that significant. We would hook into CompilationProvider and perform a Select to produce an object containing the set of types and their properties needed to generate the serializer (the minimum amount of work we can) - as that would be used for caching in the incremental pipeline we would still get the benefit of not having to execute the rest of the complex work (verification of property types, generating outputs). I think we should plan to have a conversation and dive a bit deeper into how this works in practice.

Also I found this cool lib which could be helpful https://github.com/ByronMayne/SourceGenerator.Foundations

**Comment by VaclavElias** [comment]
Another Topic to discuss - Update platform support https://github.com/stride3d/stride-website/issues/308

**Reply by Doprez** [reply]
@MeharDT it would be good to get your input on the above for mobile. I think you are the most knowledgable in recent time AFAIK.

@Jklawreszuk If you have anything to add for linux as well it would be good to hear.

For mac I dont think there is currently anyone who is using it AFAIK but I think its fully broken due to the M1 chips.

**Reply by MeharDT** [reply]
Thanks! Also tagging @Basewq as he's a lot more knowledgeable about the Silk.NET issues on mobile

**Reply by manio143** [reply]
I learned recently that M1 macs can run x64 processes via emulation. Not sure if this is usable for Stride or if graphics APIs need native execution. Some thread touching on people trying things out https://github.com/dotnet/sdk/issues/22380

**Comment by Kryptos-FR** [comment]
**What I am working on:**
* I am continuing my work on the editor rewrite.
  * Depending on finding the time I could make a video explaining what I've done and the next steps.
* I'd like to occasionally contribute on other subjects, e.g. Linux build (thanks @Jklawreszuk for the work so far).

**Comment by Eideren** [comment]
What I am working on:
- Integrating Bepu although currently on hold while Norbo looks into a critical bug.
  - Wondering whether we should move samples over to this new physics system on release.
  - What should we do with the current one ?
    - Supporting both at the same time doesn't sound super great, 
    - Asking which physics system users are using when filing issues or asking questions might be a pain
  - Would this warrant increasing minor version to 4.3/4.4 ?

**Comment by ykafia** [comment]
What I'm working on :

SDSL :
* Revamped the parser for a more performant one.
* Revamped the vscode extension to be able to run a language server. 
* Created a language server using Omnisharp's core library. 
* Currently working on the actual language and compiler design to understand where constraints exists.

SDFX :
* Parser is WIP


One thing that we could talk about is SDSL is its ties to HLSL and SPIR-V and if we can add syntax that are closer to C# rather than use the C-like syntax HLSL and GLSL use.

**Comment by Basewq** [comment]
I don't have anything new for discussion, but in case I don't make the meeting for some of the topics already raised:

- **Next Release:** The texture bug https://github.com/stride3d/stride/issues/2376 should be a priority to fix (at a minimum, revert for Windows)
- **Android:** FYI, not to bore with the details I'm no longer actively working on this area (I will still keep an eye on the issues I already raised). Sadly, the roadblock is Silk and as per https://github.com/dotnet/Silk.NET/issues/2232 progression for having Android apps on newer OS work is now at risk.

What I'm working on:
- Nothing within Stride codebase, but I am playing around with the Material shader system. I am considering writing some sort of supplement or companion guide, but the amount of explanation required keeps growing and growing...

**Reply by Perksey** [reply]
Yep we’re low on maintenance effort for 2.X atm. Work was started on the fix but we don’t have the people.

**Reply by MeharDT** [reply]
Hi @Perksey, if I'm following correctly the two outstanding issues are,

- The SDL Android project in Silk.NET 2.x needs to be updated to the latest version (the existing PR doesn't appear to fix the bug per @Basewq's description https://github.com/dotnet/Silk.NET/pull/2223)
- Once the PR is approved, the Silk.NET binaries need to be re-built with these fixes

**Reply by Perksey** [reply]
Yes this is syncing our Java code with whatever the latest is in SDL - if it works upstream, it'll work in silk. But yeah work has stalled there as there aren't any maintainers able to dedicate time to this.

**Reply by MeharDT** [reply]
The PR work seems pretty straight forward, although I've only had a quick look. My only concern if we tackle it is there would need to be a way to test the integration with Silk and bug fix on Android. The issue has been very inconsistent for me but I know some users hit it 100% of the time.

That said, it sounds like we're going to hit a roadblock with issue 2 regardless. Would anyone on the Silk.NET team be open to a bounty or sponsorship for this task? We need to fix mobile support in Stride, and from the linked issue there are other Silk projects that need this fix as well.

**Reply by Perksey** [reply]
Yes most likely.

**Reply by MeharDT** [reply]
@Basewq We should be unblocked on the Android front with the latest Silk.NET release, do you still have bandwidth/interest in the issue?

https://github.com/dotnet/Silk.NET/releases/tag/v2.22.0

We also have this open PR related to the issue, which I believe was only a partial fix?

https://github.com/stride3d/stride/pull/2351

**Reply by Basewq** [reply]
> @Basewq We should be unblocked on the Android front with the latest Silk.NET release, do you still have bandwidth/interest in the issue?

I can probably take a look later this week/next week. If you or anyone else is able update the silk version and see if there's any breaking changes, that'll be great, otherwise I'll see what happens when I get around to it.

**Comment by xen2** [comment]
Getting back to speed.
No exact plan decided yet, but I was considering helping on a few topics (one of them being the shader system rewrite with @ykafia).

**Comment by manio143** [comment]
I'm not actively working on Stride at the moment. Only sometimes checking PRs (because I'm subscribed to the repo updates through email). Please @ me on discord if you want my input on something.

**Comment by Jklawreszuk** [comment]
I'm currently not focus on anything specific too. Just waitin for resolving current PR's. Although I would like to continue my work on better Linux support, overall experience and testing future Avalonia editor if that would be need :D

**Comment by Doprez** [comment]
My focus is very mixed but I will be primarily working on the below:

- external windowing support (https://github.com/stride3d/stride/pull/2404) Currently working on creating the process of having a user managed Stride lifetime. The portion of the PR that creates the window and passes the handle seems to work as expected.

- .NET 9 branch testing (mostly to try and keep up to date with any changes that may come when the LTS version arrives) This one is mostly for fun since bigger number is exciting.

- Bepu navigation improvements (Slowly, but still working on trying to make this as user friendly and efficient as the current system) I need to merge what I have done into the main PR instead of using the previous external library but the changes should be ready to go for me AFAIK.
